subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java

```diff
@@ -0,0 +1,155 @@
+
+import org.gradle.api.Incubating;
+import org.gradle.tooling.exceptions.MultipleBuildActionsException;
+
+/**
+ * Used to execute multiple {@link BuildAction}s in different phases of the build process.
+ *
+ * @since 4.7
+ */
+@Incubating
+public interface PhasedBuildActionExecuter extends ConfigurableLauncher<PhasedBuildActionExecuter> {
+
+    /**
+     * Builder for a {@link PhasedBuildActionExecuter}.
+     *
+     * <p>A single {@link BuildAction} is allowed per build phase. Use composite actions if needed.
+     *
+     * @since 4.7
+     */
+    @Incubating
+    interface Builder {
+
+        /**
+         * Adds the given action and its result handler to be executed after projects are loaded.
+         *
+         * <p>The action or model builders invoked by it are run when plugins have not yet be applied, and so it should try to get only gradle default models.
+         *
+         * <p>If the operation fails, the handler's {@link ResultHandler#onFailure(GradleConnectionException)} method is called with the appropriate exception.
+         *
+         * @param buildAction The action to run in the specified build phase.
+         * @param handler The handler to supply the result of the given action to.
+         * @param <T> The returning type of the action.
+         * @return The builder.
+         * @throws MultipleBuildActionsException If an action has already been added to this build phase. Multiple actions per phase are not supported yet.
+         *
+         * @since 4.7
+         */
+        <T> Builder addAfterLoadingAction(BuildAction<T> buildAction, ResultHandler<? super T> handler) throws MultipleBuildActionsException;
+
+        /**
+         * Adds the given action and its result handler to be executed after projects are configured and before tasks are run.
+         *
+         * <p>The action has already access to plugins' model builders. These model builders can modify the task graph, since tasks have not yet been run.
+         *
+         * <p>If the operation fails, the handler's {@link ResultHandler#onFailure(GradleConnectionException)} method is called with the appropriate exception.
+         *
+         * @param buildAction The action to run in the specified build phase.
+         * @param handler The handler to supply the result of the given action to.
+         * @param <T> The returning type of the action.
+         * @return The builder.
+         * @throws MultipleBuildActionsException If an action has already been added to this build phase. Multiple actions per phase are not supported yet.
+         *
+         * @since 4.7
+         */
+        <T> Builder addAfterConfigurationAction(BuildAction<T> buildAction, ResultHandler<? super T> handler) throws MultipleBuildActionsException;
+
+        /**
+         * Adds the given action and its result handler to be executed after tasks are run.
+         *
+         * <p>The action should not invoke task graph modifiers, since tasks have already been run.
```

Comment:
⭕️ "task graph modifier" is not a term we use. I think I would leave this out. It's also kind of obvious from the fact that this is called "buildFinished". I don't think anyone would expect a modification of the StartParameter to have any effect at that point.